### PR TITLE
Fix mutex use after free in llvm_shutdown

### PR DESCRIPTION
**Pull request by expipiplus1**
This patch is consolidated from the following upstream patches:

- llvm/llvm-project@3d5360a4398b
- llvm/llvm-project@928071ae4ef5
- llvm/llvm-project@56349e8b6d85

Fixes https://github.com/microsoft/DirectXShaderCompiler/issues/5119

### CODE REVIEW

File: include/llvm/Support/ManagedStatic.h [hunk 1]
```diff
@@ -1,9 +1,8 @@
 //===-- llvm/Support/ManagedStatic.h - Static Global wrapper ----*- C++ -*-===//
 //
-//                     The LLVM Compiler Infrastructure
-//
-// This file is distributed under the University of Illinois Open Source
-// License. See LICENSE.TXT for details.
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
```
Comment (llvm-beanz) [inline include/llvm/Support/ManagedStatic.h:3]: I appreciate that this is coming from a cherry-pick, but changing the file license is... not something we can do in a cherry-pick.

LLVM's code is currently dual licensed, so it is still also available under the UIUC open source license which is what DXC's code is mostly under.

File: include/llvm/Support/ManagedStatic.h [hunk 2]
```diff
@@ -1,9 +1,8 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
 //===----------------------------------------------------------------------===//
 //
```
Comment (llvm-beanz) [inline include/llvm/Support/ManagedStatic.h:8]: Can you add a comment near the top of the file (after the header) listing the LLVM commits that you've back-ported. That will make it easier for us to see from just the source where these changes are coming from.

File: lib/Support/ManagedStatic.cpp [hunk 3]
```diff
@@ -1,9 +1,8 @@
 //===-- ManagedStatic.cpp - Static Global wrapper -------------------------===//
 //
-//                     The LLVM Compiler Infrastructure
-//
-// This file is distributed under the University of Illinois Open Source
-// License. See LICENSE.TXT for details.
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```
Comment (llvm-beanz) [inline lib/Support/ManagedStatic.cpp:5]: Same issue here with the license.

File: lib/Support/ManagedStatic.cpp [hunk 4]
```diff
@@ -1,9 +1,8 @@
...
 //
 //===----------------------------------------------------------------------===//
 //
```
Comment (llvm-beanz) [inline lib/Support/ManagedStatic.cpp:8]: Also please add a comment near the top of this file listening the changes that you're back porting.

**Comment by llvm-beanz** [comment]
Tagging for our upcoming release since I think this would be nice to get in.

**Comment by expipiplus1** [comment]
@llvm-beanz, Thanks, I've resolved your comments.

Since these are just picks from upstream I don't think the CLA is appropriate for me to sign here. Can I leave this to you from here, please, to reproduce or merge without.

**Comment by llvm-beanz** [comment]
> @llvm-beanz, Thanks, I've resolved your comments.
> 
> Since these are just picks from upstream I don't think the CLA is appropriate for me to sign here. Can I leave this to you from here, please, to reproduce or merge without.

We can’t bypass the CLA bot, and it does still apply for contributions that are not material you own the copyright on.

We haven’t been strict about the specific phrasing the CLA lists (because we never really remember it), but compliance with the CLA is why we (as you did) list the upstream commits in the description, and sometimes in comments in the source.

**Review by llvm-beanz — APPROVED** [review]

**Comment by expipiplus1** [comment]
The bot's message has disappeared, but should it come back I'll read it again with that in mind.

**Comment by llvm-beanz** [comment]
> The bot's message has disappeared, but should it come back I'll read it again with that in mind.

Yea, not sure what is going on with the CLA bot ☹️ 

The CLA document is linked from [this page](https://opensource.microsoft.com/cla/).
The CLA bot commands are documented [here](https://github.com/microsoft/ContributorLicenseAgreement).

**Comment by expipiplus1** [comment]
@microsoft-github-policy-service agree

**Review by python3kgae — APPROVED** [review]

**Pull request merged**